### PR TITLE
Get all clients in a room across all nodes

### PR DESCRIPTION
**Pull request by fgodino**
This pull request adds the ability to get all clients IDs connected to a room or rooms. It's up to date with `Adapter` so a call to [Namespace#clients(fn:Function)](https://github.com/socketio/socket.io#namespaceclientsfnfunction) will return the same result as `RedisAdapter.clients(rooms:Array, fn:Function)`

### CODE REVIEW

**Comment by artofspeed** [comment]
can i safely use `socket.rooms` to get all rooms a socket has joined?

**Comment by artofspeed** [comment]
Also, `.clients` function seems to be undefined.

```
var socketRedis = require('socket.io-redis');
var adapter = socketRedis({ host: 'localhost', port: 6380 });
var io = socketio(server);
io.adapter(adapter);

console.log(adapter.clients); // undefined
console.log(io.of('/').clients); // undefined

```

**Comment by fgodino** [comment]
@artofspeed, socket.rooms is noy implemented although it would not require a lot of work. Regarding the incompatibility with v1.3.7, you can access this function via `adapter` attribute (haven't tried). If you need a `rooms` functionality I can add it to this PR

**Comment by dakairus** [comment]
Is there any intentions soon by the maintainer to implement this PR? is a great feature that I guess a lot of people is wanting as me.

**Comment by naholyr** [comment]
+1 this is a really lacking feature in current implementation

**Comment by bloodfire91** [comment]
any timeline on when this will be merged?

**Comment by darrachequesne** [comment]
Hi @fgodino ! Thanks for your PR. I'll try to test that ASAP.

Is anybody already using it? Any shortcomings?

**Comment by fgodino** [comment]
Hey @darrachequesne, we are already using it in our website https://www.varsityviews.com/collections/ and no problems so far.

**Comment by hardeep-proptiger** [comment]
Hey @fgodino,
I have tried to implement it, Although clients is a known function, But rooms paramater is coming out to be undefined.

Any solution?

File: index.js [hunk 1]
```diff
@@ -125,6 +140,41 @@ function adapter(uri, opts){
...
+  Redis.prototype.onclients = function(channel, msg){
+
+    if (!this.channelMatches(channel.toString(), this.syncChannel)) {
+      return debug('ignore different channel');
+    }
+
+    try {
+      var decoded = JSON.parse(msg);
+    } catch(err){
+      self.emit('error', err);
```
Comment (darrachequesne) [inline index.js]: Note: `self` seems undefined here.

File: index.js [hunk 2]
```diff
@@ -236,10 +286,88 @@ function adapter(uri, opts){
...
+        if (fn) fn(err);
+        return;
+      }
+
+      numsub = numsub[1];
+
+      var msg_count = 0;
+      var clients = {};
+
+      subJson.on("subscribe", function subscribed(channel, count) {
```
Comment (darrachequesne) [inline index.js]: Any particular reason for double quotes here?

File: index.js [hunk 3]
```diff
@@ -236,10 +286,88 @@ function adapter(uri, opts){
...
+
+        var request = JSON.stringify({
+          transaction : transaction,
+          rooms : rooms
+        });
+
+        /*If there is no response for 5 seconds, return result;*/
```
Comment (darrachequesne) [inline index.js]: 5 seconds?

File: index.js [hunk 4]
```diff
@@ -236,10 +286,88 @@ function adapter(uri, opts){
...
+
+        subJson.on(subEvent, function onEvent(channel, msg) {
+
+          if (!self.channelMatches(channel.toString(), responseChn)) {
+            return debug('ignore different channel');
+          }
+
+          var response = JSON.parse(msg);
+          
+          for(var i = 0; i < response.clients.length; i++){
```
Comment (darrachequesne) [inline index.js]: Could you maybe add a test for the existence of the `clients` property?

**Comment by darrachequesne** [comment]
@artofspeed a socket should only be connected to one node at a time, so `socket.rooms` should be safe.

Besides, I think `socket.io-redis` v1.0.0 is already incompatible with `socket.io` v1.3.7, so that should not matter much :angel:

File: index.js (hunk 1 above)
Comment (fgodino) [inline index.js]: Fixed

**Comment by darrachequesne** [comment]
@fgodino thanks a lot!

**Comment by fgodino** [comment]
@darrachequesne np!

**Comment by hems** [comment]
the funny thing is: if a user opens multiple tabs and go to different rooms in multiple tabs it might get a different socket.id in each of the connections and then you might not be able to know all the different socket.ids represent the same user?

Does that make sense on your application context? On my app every time a user logs in ( for instance using facebook or chrome ) i have to save his socket.id to the database, then if the user open another tab he will have two socket.id ( one in each tab ) and things start to get a bit complicated... It's manageable, but not exactly "programmer friendly".. Am i doing something wrong? It sounds like there must be an easier way.

**Comment by darrachequesne** [comment]
@hems Managing users (and not sockets) is out of the scope of that library (what is a user is very much context-dependant)

Nonetheless, couldn't you create a room dedicated to each user, and make every socket join that room once authenticated?

**Comment by hems** [comment]
@darrachequesne yeah it makes sense. Still i have to keep a list of socket.ids a logged user has so when retrieving the list of users from a room i can find the user profile on my database and display for instance his name/avatar.

**Comment by artofspeed** [comment]
@darrachequesne @fgodino is it possible to get all rooms a socket has joined, given the `socket id` only? `socket.rooms` only works when executed on the node instance that contains that `socket object`.

**Pull request merged**